library/core/src/ffi/c_str.rs

```diff
@@ -939,4 +990,75 @@ mod verify {
         assert_eq!(bytes, &slice[..end_idx]);
         assert!(c_str.is_safe());
     }
-}
\ No newline at end of file
+
+    // Proof harness that uses the const-evaluated length
+    // #[kani::proof_for_contract(strlen)]
+    #[kani::proof]
+    #[kani::unwind(32)]
+    fn check_strlen_const_path() {
+        // Get the compile-time computed length
+        const BYTES: &[u8] = b"test\0";
+        const PTR: *const c_char = BYTES.as_ptr() as *const c_char;
+        
+        // This must be evaluated at compile time
+        const COMPUTED_LEN: usize = unsafe { strlen(PTR) };
+        
+        // Verify the const-computed length
+        assert_eq!(COMPUTED_LEN, 4);
+        
+        // Additional compile-time verifications
+        const BYTES1: &[u8] = b"test\0";
+        const PTR1: *const c_char = BYTES1.as_ptr() as *const c_char;
+        
+        // Verify the string properties using the const-computed length
+        unsafe {
+            // These assertions will use the const-computed values
+            assert_eq!(*PTR1.add(COMPUTED_LEN), 0);
+            
+            let mut i = 0;
+            while i < COMPUTED_LEN {
+                assert_ne!(*PTR1.add(i), 0);
+                i += 1;
+            }
+        }
+    }
+
+    // Stub for external C strlen
+    #[allow(dead_code)]
+    mod stubs {
+        use super::*;
```

Comment:
Based on my research on stubs : 
https://github.com/model-checking/kani/issues/2423
https://github.com/firecracker-microvm/firecracker/blob/dbd9a84b11a63b5e5bf201e244fe83f0bc76792a/src/rate_limiter/src/lib.rs#L607-L639
https://github.com/model-checking/kani/pull/2658

I figured we can stub the external strlen function individually and write a proof for it, But this seems to not work as expected and gives me this error.

```
   Compiling dummy v0.1.0 (/Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/target/kani_verify_std)
error: Failed to find contract closure `__kani_recursion_check_strlen` in function `ffi::c_str::strlen`

error: could not compile `core` (lib) due to 1 previous error
error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
```

Reply:
The problem here is that the strlen you are stubbing is the one you are trying to check.

There's no way for Kani to stub a nested function. Can you pass `-Z c-ffi` to Kani and remove the stub to see if that would work? Otherwise, we can add support for strlen to Kani. I believe CBMC already has a model available.

Reply:
Hi @celinval , thank you, using the flag seems to work, I have made some changes to the code and pushed it. Pl review